pages/data-providers/data-provider.jsx

```diff
@@ -1,7 +1,102 @@
-import React from 'react'
+import React, { useEffect, useContext, useRef } from 'react'
+import { useRouter } from 'next/router'
 
+import { withGlobalStore, GlobalContext } from 'store'
 import DataProviderPageTemplate from 'templates/data-provider'
+import useDebouncedEffect from 'hooks/use-debounced-effect'
 
-const DataProviderPage = () => <DataProviderPageTemplate />
+// TODO: Pages should be just a thin layer of NextJS.
+//       Find a better place for it.
+const useDataProviderController = () => {
+  const formRef = useRef(null)
+  const router = useRouter()
+  const { dataProvider } = useContext(GlobalContext)
 
-export default DataProviderPage
+  useDebouncedEffect(
+    () => {
+      if (dataProvider.query === '') return
+      if (!formRef.current?.dataProviderUrl.checkValidity()) {
+        formRef.current.dataProviderUrl.reportValidity()
+        return
+      }
+      dataProvider.retrieve()
+    },
+    [dataProvider.query],
+    500
+  )
+
+  useEffect(() => {
+    if (router.query?.dataProviderUrl)
+      dataProvider.query = router.query?.dataProviderUrl
+  }, [router.query?.dataProviderUrl])
```

Comment:
Shall we update this query every time the user changes the input string?

Reply:
I am not sure about that. Let me think about it.

Reply:
I think this is a good feature but we should not update it immediately but only before the request to the API. It will allow us to have sharable and percistent pages (with predictable API responses from the user point of view). We can potentially use it to check what the user sees.

This is not a requirement. Let's keep this in mind 🙂